anndata/_core/anndata.py

```diff
@@ -783,12 +830,12 @@ def raw(self):
     @property
     def n_obs(self) -> int:
         """Number of observations."""
-        return self._n_obs
+        return len(self.obs)
```

Comment:
Is this critical to the PR? 

* I can forsee issues with delayed/ out of core objects.
* I think the `len` of a dataframe is not intuitive

```python
import pandas as pd

df = pd.DataFrame({"a": [1, 2]})

assert len([i for i in df]) != len(df)
```

* I am eyeing moving towards the dataframe API, which does not define `__len__`
* It would make sense if `obs` and `var` were also views, but then `len(self.obs)` would cause compute.

As an alternative:

```suggestion
        return len(self.obs_names)
```

Since I expect those will always be around.